### PR TITLE
[SPARK-43032][CONNECT][SS] Add Streaming query manager

### PR DESCRIPTION
**Pull request by WweiL**
### What changes were proposed in this pull request?

Add support of `StreamingQueryManager()` to CONNECT PYTHON client.

### Why are the changes needed?

Now users can use typical streaming query manager method by calling `session.streams`

### Does this PR introduce _any_ user-facing change?

Yes

### How was this patch tested?

Manual test and unit test
```
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /__ / .__/\_,_/_/ /_/\_\   version 3.5.0.dev0
      /_/

Using Python version 3.9.16 (main, Dec  7 2022 01:11:58)
Client connected to the Spark Connect server at localhost
SparkSession available as 'spark'.
>>> q = spark.readStream.format("rate").load().writeStream.format("memory").queryName("test").start()
23/04/19 23:10:43 WARN ResolveWriteToStream: Temporary checkpoint location created which is deleted normally when the query didn't fail: /tmp/temporary-913e48b9-26d8-448f-899f-d9f5ae08707d. If it's required to delete it under any circumstances, please set spark.sql.streaming.forceDeleteTempCheckpointLocation to true. Important to know deleting temp checkpoint folder is best effort.
23/04/19 23:10:43 WARN ResolveWriteToStream: spark.sql.adaptive.enabled is not supported in streaming DataFrames/Datasets and will be disabled.
>>> spark.streams.active
[<pyspark.sql.connect.streaming.query.StreamingQuery object at 0x7f4590400d90>]
>>> q1 = spark.streams.active[0]
>>> q1.id == q.id
True
>>> q1.runId == q.runId
True
>>> q1.runId == q.runId
True
>>> q.name
'test'
>>> q1.name
'test'
>>> q == q1
False
>>> q1.stop()
>>> q
<pyspark.sql.connect.streaming.query.StreamingQuery object at 0x7f4590400b20>
>>> q1
<pyspark.sql.connect.streaming.query.StreamingQuery object at 0x7f4590400ee0>
>>> q.isActive
False
```

### CODE REVIEW

File: connector/connect/common/src/main/protobuf/spark/connect/base.proto [hunk 1]
```diff
@@ -304,8 +304,11 @@ message ExecutePlanResponse {
     // Response for commands on a streaming query.
     StreamingQueryCommandResult streaming_query_command_result = 9;
 
+    // Response for commands on the streaming query manager.
+    StreamingQueryManagerCommandResult streaming_query_manager_command_result = 10;
+
     // Response for 'SparkContext.resources'.
-    GetResourcesCommandResult get_resources_command_result = 10;
+    GetResourcesCommandResult get_resources_command_result = 11;
```
Comment (amaliujia) [inline connector/connect/common/src/main/protobuf/spark/connect/base.proto:311]: This is a breaking change... 

probably you just do `StreamingQueryManagerCommandResult streaming_query_manager_command_result = 11;` and do not re-use existing proto number

Comment (WweiL) [reply connector/connect/common/src/main/protobuf/spark/connect/base.proto:311]: I see. TY i'll change it back
Just wanted to make similar names closer...

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto [hunk 2]
```diff
@@ -236,6 +237,9 @@ message StreamingQueryInstanceId {
   // will generate a unique run_id. Therefore, every time a query is restarted from
   // checkpoint, it will have the same `id` but different `run_id`s.
   string run_id = 2;
+
+  // (Optional) The name of this query.
+  optional string name = 3;
```
Comment (rangadi) [inline connector/connect/common/src/main/protobuf/spark/connect/commands.proto:242]: Why is this needed. Note that query-name is not part of real identity of the query.  This can be extra field like response for 'WriteStreamOperation'.

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto [hunk 3]
```diff
@@ -321,6 +324,50 @@ message StreamingQueryCommandResult {
...
+message StreamingQueryManagerCommandResult {
+  oneof result_type {
+    ActiveResult active = 1;
+    StreamingQueryInstanceId query = 2;
+    AwaitAnyTerminationResult await_any_termination = 3;
+    bool reset_terminated = 4;
+  }
+
+  message ActiveResult {
+    repeated StreamingQueryInstanceId active_queries = 1;
```
Comment (rangadi) [inline connector/connect/common/src/main/protobuf/spark/connect/commands.proto:363]: This can be the oneof field.

**Review by rangadi — COMMENTED** [review]
Mostly LGTM. Left a couple of comments.

**Comment by WweiL** [comment]
PTAL! @amaliujia @HyukjinKwon Thank you!

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto (hunk 3 above)
Comment (WweiL) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:363]: Yes that's my original thought. But I got IntelliJ warning `Oneof fields cannot have labels` so that's why I ended up doing it like this

Comment (rangadi) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:363]: Oh, that might be. Does it fail to compile. We can keep as you have now.

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto (hunk 2 above)
Comment (WweiL) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:242]: Right I'm also not very sure here. Basically I added this because here:
https://github.com/apache/spark/blob/069caf4dcdb4380cd6eeca394b55824d5b71940a/connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala#L2411
In the `sqm.get(queryID)` handler, the query may or may not have a name. But in case it has, if we don't return the name then the client won't have it's name.

We could also maintain a local cache of query names, but that would add more complexity on cleaning up that cache though... What do you think

File: python/pyspark/sql/connect/client.py [hunk 4]
```diff
@@ -967,9 +967,8 @@ def _execute_and_fetch_as_iterator(
                                 "streaming_query_command_result": b.streaming_query_command_result
                             }
                         if b.HasField("streaming_query_manager_command_result"):
-                            yield {
-                                "streaming_query_manager_command_result": b.streaming_query_manager_command_result
-                            }
+                            cmd_result = b.streaming_query_manager_command_result
```
Comment (WweiL) [inline python/pyspark/sql/connect/client.py:970]: If I don't do this, `dev/reformat-python` always changes this line to
```
yield {
                                "streaming_query_manager_command_result": b.streaming_query_manager_command_result
                            }
```
which would result to python lint error because the second line has length 115

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto (hunk 2 above)
Comment (rangadi) [reply connector/connect/common/src/main/protobuf/spark/connect/commands.proto:242]: We can return the name, but we don't need to change `StreamingQueryInstanceId`. We can return both `StreamingQueryInstanceId` and the name

File: python/pyspark/sql/connect/client.py (hunk 4 above)
Comment (rangadi) [reply python/pyspark/sql/connect/client.py:970]: Looks fine to me.

File: python/pyspark/sql/tests/connect/streaming/test_parity_streaming.py [hunk 5]
```diff
@@ -22,32 +22,7 @@
...
 class StreamingParityTests(StreamingTestsMixin, ReusedConnectTestCase):
-    @unittest.skip("Query manager API will be supported later with SPARK-43032.")
-    def test_stream_await_termination(self):
-        super().test_stream_await_termination()
-
-    @unittest.skip(
-        "Query immediately quits after throw, "
-        + "allowing access to supported queries will be added in SPARK-42962."
-    )
-    def test_stream_exception(self):
```
Comment (rangadi) [inline python/pyspark/sql/tests/connect/streaming/test_parity_streaming.py:33]: Do you know why this was here before?

File: python/pyspark/sql/tests/connect/test_parity_pandas_grouped_map_with_state.py [hunk 6]
```diff
@@ -25,10 +25,6 @@
 class GroupedApplyInPandasWithStateTests(
     GroupedApplyInPandasWithStateTestsMixin, ReusedConnectTestCase
 ):
-    @unittest.skip("Query manager API will be supported later with SPARK-43032.")
-    def test_apply_in_pandas_with_state_basic(self):
-        super().test_apply_in_pandas_with_state_basic()
-
     @unittest.skip("Query manager API will be supported later with SPARK-43032.")
```
Comment (rangadi) [inline python/pyspark/sql/tests/connect/test_parity_pandas_grouped_map_with_state.py:28]: This one needs to be fixed too?

Comment (WweiL) [reply python/pyspark/sql/tests/connect/test_parity_pandas_grouped_map_with_state.py:28]: This one calls foreachBatch

Comment (rangadi) [reply python/pyspark/sql/tests/connect/test_parity_pandas_grouped_map_with_state.py:28]: Lets update the skip reason then.

File: python/pyspark/sql/tests/connect/streaming/test_parity_streaming.py (hunk 5 above)
Comment (WweiL) [reply python/pyspark/sql/tests/connect/streaming/test_parity_streaming.py:33]: Yes because it access stopped query. We seem to have a slight conflict with spark tasks (42962 and the real better session mgmt one). Anyway it should be fine after this PR

**Review by HyukjinKwon — APPROVED** [review]

File: connector/connect/common/src/main/protobuf/spark/connect/commands.proto [hunk 7]
```diff
@@ -238,15 +239,21 @@ message StreamingQueryInstanceId {
   string run_id = 2;
 }
 
+message StreamingQueryInstance {
```
Comment (rangadi) [inline connector/connect/common/src/main/protobuf/spark/connect/commands.proto:242]: Add a short comment. This name and `StreamingQueryInstanceId` might be confusing.
May be we should remove the `name` field completely and make it an API call to server when it is needed (in a follow up). 
For now, could you move it inside `StreamingQueryManagerCommand` as 'GetQueryResult`?

**Review by rangadi — APPROVED** [review]
LGTM. Suggested one change, but that is ok, we could merge this.

**Comment by WweiL** [comment]
Also verified null query names:
```
>>> q = spark.readStream.format("rate").load().writeStream.format("console").start()
>>> q1 = spark.streams.get(q.id)
>>> q1.name
''
>>> q.name
''
>>> q.name == q1.name
True
```

**Review by rangadi — APPROVED** [review]

**Comment by xinrong-meng** [comment]
Merged to master, thanks!
